### ISSUE TITLE
flambda with -nostdlib not finding cmx

### ISSUE THREAD

**Issue by EduardoRFS** [ocaml/dune#4039]
## Expected Behavior
Dune should point the cmx folder when building the executable

## Actual Behavior

```
Warning 58: no cmx file was found in path for module Stdlib
```

## Reproduction

The README explains how to try it, but is essentially make a flambda switch and try to compile with 

https://github.com/EduardoRFS/dune-flambda-nostdlib-no-cmx

## Specifications

- Version of `dune` (output of `dune --version`): 2.7.1
- Version of `ocaml` (output of `ocamlc --version`): 4.09.1
- Operating system (distribution and version): Arch Linux
- Link to gist with verbose output (run `dune` with the `--verbose` flag): https://github.com/EduardoRFS/dune-flambda-nostdlib-no-cmx/blob/main/build.output.txt

**Comment by nojb** [comment]
You are passing `-nostdlib`. Could you clarify what behaviour would you like in this situation?

**Comment by EduardoRFS** [comment]
@nojb I'm passing `-nostdlib` for a library, as you can see in the example, `lib_a` is using `-nostdlib` but `custom_stdlib` isn't, and `lib_a` depends on `custom_stdlib` so while `lib_a` shouldn't have access to the stdlib `.cmi` it should have access to the stdlib `.cmx`

The problem goes away as soon as the executable has a direct reference for the stdlib on the executable.

**Comment by EduardoRFS** [comment]
I also updated the description to include more detail, but the reproduction is simple, try it with closure vs flambda and you're going to notice the problem.

**Comment by nojb** [comment]
If I'm understanding correctly, you would like to make `stdlib.cmx` visible to the compiler but not `stdlib.cmi`. This is not currently possible to do, and would need some work at the compiler level to achieve.

**Comment by EduardoRFS** [comment]
@nojb stdlib.cmx is only is only needed for optimizations during the linking of an executable, it's not needed to build the actual module. That's the problem. There is no problem to add the full folder during linking, as `.cmi` will not even be used

**Comment by nojb** [comment]
Yes, the `.cmx` is used during compilation to perform inlining. Perhaps `dune` could help in some way, but that would require `dune` having a higher-level understanding of what you are trying to achieve using `-nostdlib`.

For your use-case in particular, couldn't you play with shadowing to achieve an equivalent effect, eg by opening a module with a top-level `Stdlib` submodule?

**Comment by EduardoRFS** [comment]
@nojb I will give some tries, maybe we can still get the optimization with some dune rules to copy things. That seems to be a big problem if someone wants to use `(implicit_transitive_deps false)`

**Comment by EduardoRFS** [comment]
Tried `(implicit_transitive_deps false)` on a private project and it's exactly that, a lot of inlining oportunities are lost because there is `no cmx file available`.

Seems like the compiler needs a `-L`.

**Comment by mobileink** [comment]
> Yes, the `.cmx` is used during compilation to perform inlining. 

Are you sure about that?  I would expect such optimizations to occur only at link time. If they're done during compilation you'd end up repeating the same optimization.

If the .cmx is missing the compiler just issues a warning, which I take to mean: "you have not provided a .cmx for this guy, but you have also not disabled cross-module optimization (you did not pass -opaque), so if you're expecting cross-module optimization at link time you're probably going to be disappointed."

But a clever build system should be able to omit cmx deps at compile time while providing them at link time.
